CONTRIBUTING.md

```diff
@@ -173,6 +173,47 @@ The `legacy_foo` arg was removed
 :::
 ```
 
+## Style and idioms
+
+For the most part, we just accept whatever the code formatters do, so there
+isn't much style to enforce.
+
+Some miscellanous style, idioms, and conventions we have are:
+
+### Markdown/Sphinx Style
+
+* Use colons for prose sections of text, e.g. `:::{note}`, not backticks.
+* Use backticks for code blocks.
+* Max line length: 100.
+
+### BUILD/bzl Style
+
+* When a macro generates public targets, use a dot (`.`) to separate the
+  user-provided name from the generted name. e.g. `foo(name="x")` generates
+  `x.test`. The `.` is our convention to communicate that it's a generated
+  target, and thus one should look for `name="x"` when searching for the
+  definition.
+* The different build phases and interface/implementation pieces of rules should
+  be in different files. Stated another way: loading code for thing shouldn't
+  incur loading code that isn't relevant.
+  * Providers should be in their own files. This allows implementing a custom
+    rule that implements the provider without loading an existing
+    implementation.
+  * The repository phase shouldn't `load()` code that is only relevant to the
+    loading-phase or subsequent phases. Vice-versa, loading-phase code should
+    not load any repository-phase code.
```

Comment:
What is the rationale here, e.g. we have `normalize_name` function that is used in both, repository and loading phases.

Or am I misunderstanding the statement here?

Reply:
I'll rephrase it to clarify.

The basic idea is:
* The repo phase shouldn't load code that calls rule(), module_extension(), provider(), etc. The objects those define aren't usable/inspectable/valid by/to the repo phase. Vice versa: the loading phase shouldn't call code that calls repository_rule, module_extension, etc
* Similarly, utility code like repo_utils.bzl shouldn't be loaded by the loading-phase because 99% of it is only relevant to the repository phase.

"generic" code (e.g. normalize_name) is fine for any phase to load.

Reply:
SGTM, thanks